### PR TITLE
[#147446215] Basic integration test of mongo service

### PR DESCRIPTION
**Pull request by 46bit**
## What

As described [in the story](https://www.pivotaltracker.com/story/show/147446215) we want confidence that the Compose broker (for MongoDB) works in a real Cloud Foundry environment.

We have added MongoDB to the existing `platform-tests/example-apps/healthcheck` app. This is used to perform simple operations on databases to prove their basic functionality works. This will now support MySQL, Postgres and now MongoDB.

We have then added acceptance tests to `platform-tests/src/acceptance/compose_broker_test.go` that deploy the `healthcheck` application and test that it successfully performs read-and-write operations to MongoDB.

The commit descriptions provide a further guide to what has been done.

## How to review

The new acceptance tests can be run by:

1. Executing `BRANCH=feature/147446215-mongodb-integration-test make dev pipelines` on `paas-cf`
2. Rerunning the `create-cloudfoundry` pipeline. Our changes affect the `custom-acceptance-tests` task.

## Who can review

Not @dcarley @timmow @46bit.

### CODE REVIEW

File: platform-tests/example-apps/healthcheck/db.go [hunk 1]
```diff
@@ -105,3 +129,94 @@ func mysqlOpen(dbu string, ssl bool) (*sql.DB, error) {
...
+	var vcap_services VCAPServices
+	if err := json.Unmarshal([]byte(vcap_services_str), &vcap_services); err != nil {
+		return nil, err
+	}
+	mongodb_credentials := vcap_services.MongoDB[0].Credentials
+	dbu := mongodb_credentials.URI
+	ca_certificate_base64 := mongodb_credentials.CACertificateBase64
+
+	// This is work around for https://github.com/go-mgo/mgo/issues/84
+	uri := strings.TrimSuffix(dbu, "?ssl=true")
```
Comment (henrytk) [inline platform-tests/example-apps/healthcheck/db.go]: I know this example is from an official source, but I've lost count of the amount of times string splitting on URLs has come back to bite me. I prefer using mature, tested libraries like the `net/url` for such things: https://play.golang.org/p/JnmoeJs8E8

File: platform-tests/example-apps/healthcheck/db.go [hunk 2]
```diff
@@ -105,3 +129,94 @@ func mysqlOpen(dbu string, ssl bool) (*sql.DB, error) {
...
+
+	if !ssl {
+		session, err := mgo.DialWithTimeout(uri, 5 * time.Second)
+		if err != nil {
+			return nil, err
+		}
+		return session, nil
+	}
+
+	// Compose has self-signed certs for mongo. Make sure we verify it against CA certificate brovided in binding
```
Comment (henrytk) [inline platform-tests/example-apps/healthcheck/db.go]: brovided
    _conjunction_

    provided by a bro

I'm adding this to Urban Dictionary.

**Comment by henrytk** [comment]
This looks good to me. Personally, I would prefer the commit history to be squashed; we don't need to know an app was used and then removed before the story finished. If you want a permanent record of the app we can fork it and stick it in [the attic](https://github.com/gds-attic).

**Comment by 46bit** [comment]
Cheers @henrytk. I'm running the pipelines with those issues you brovided and then I'll look at what we can do about squishing the app away.

File: platform-tests/src/acceptance/compose_broker_test.go [hunk 3]
```diff
@@ -1,22 +1,114 @@
...
+	"github.com/cloudfoundry-incubator/cf-test-helpers/helpers"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	. "github.com/onsi/gomega/gbytes"
 	. "github.com/onsi/gomega/gexec"
 )
 
-var _ = Describe("ComposeBroker", func() {
-
+var _ = FDescribe("Compose broker - MongoDB", func() {
```
Comment (timmow) [inline platform-tests/src/acceptance/compose_broker_test.go]: Should this be focused?

File: platform-tests/src/acceptance/compose_broker_test.go [hunk 4]
```diff
@@ -1,22 +1,114 @@
...
+			Expect(err).NotTo(HaveOccurred())
+			body, err = ioutil.ReadAll(resp.Body)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(resp.StatusCode).To(Equal(500), "Got %d response from healthcheck app. Response body:\n%s\n", resp.StatusCode, string(body))
+			Expect(string(body)).To(ContainSubstring("no reachable servers"), "Connection without TLS did not report a connection error")
+		})
 	})
 })
+
+func pollForMongoDbCreationCompletion(dbInstanceName string) {
```
Comment (henrytk) [inline platform-tests/src/acceptance/compose_broker_test.go]: These two polling functions are the same as the RDS ones with a different name; given that they are in the same package we can reduce the duplication by renaming them to be more generic. You may also opt to split them out into a separate file for helpers/utilities.

**Comment by 46bit** [comment]
@henrytk Thanks for the reviewing so far. I've squashed away the example application, done the refactoring @timmow spotted and gotten it green again on a deployment.

File: platform-tests/example-apps/healthcheck/Godeps/Godeps.json [hunk 5]
```diff
@@ -1,10 +1,7 @@
 {
-	"ImportPath": "healthcheck",
+	"ImportPath": "github.com/alphagov/paas-cf/platform-tests/example-apps/healthcheck",
```
Comment (henrytk) [inline platform-tests/example-apps/healthcheck/Godeps/Godeps.json]: You're theory seems to be right, when I moved paas-cf under my GOPATH it changed it for me too. What is a bigger mystery is why the `Packages` property on lines 5-7 was removed. Was that done manually? I don't think these changes will break anything, so happy to keep them in, but they may get changed back in future depending on who's doing the vendoring.

File: platform-tests/src/acceptance/compose_broker_test.go [hunk 6]
```diff
@@ -1,22 +1,91 @@
...
+		AfterEach(func() {
+			cf.Cf("delete", appName, "-f").Wait(DEFAULT_TIMEOUT)
+
+			Expect(cf.Cf("delete-service", dbInstanceName, "-f").Wait(DEFAULT_TIMEOUT)).To(Exit(0))
+
+			// Poll until destruction is complete, otherwise the org cleanup (in AfterSuite) fails.
+			pollForServiceDeletionCompletion(dbInstanceName)
+		})
+
+		It("binds a DB instance to mongodb-example-app that matches our criteria", func() {
```
Comment (chrisfarms) [inline platform-tests/src/acceptance/compose_broker_test.go]: "mongodb-example-app" mentioned in the test title. Think this should now be "healthcheck app"

Comment (henrytk) [inline platform-tests/src/acceptance/compose_broker_test.go]: I will push a quick fix to this @46bit

**Pull request merged**